### PR TITLE
Change responses to use classes instead of interfaces

### LINKED ISSUE #399: Change Frontend Responses to use Classes
Currently, the frontend uses a reactive system to alert different parts of the frontend about editor changes. In order to do this, all responses are described by interfaces and have a corresponding new function.
```ts
export interface SaveDocument {
	document: string;
	name: string;
}
function newSaveDocument(input: any): SaveDocument {
	return {
		document: input.document,
		name: input.name,
	};
}
```
Switching to a class-based system would have various benefits
- Classes stay in after compilation. This would allow us to offer better type checking on the main subscription API
```ts
function registerResponseHandler(responseType: ResponseType, callback: ResponseCallback)
```
to
```ts
function subscribeJsMessage<T extends JsMessage>(responseType: Constructs<T>, callback: JsMessageCallback<T>)
```
The advantage of this change is that the callback given to subscribeJsMessage automatically has the correct type for the input parameter and does not need to be cast. Removing this cuts down on boilerplate code where mistakes can be made.

- The other advantage of a class-based system is we can remove the new function to create a given type. Instead, we can make use of a library [class-transformer](https://github.com/typestack/class-transformer) in order to automate this process, further cutting down on code duplication.

- Finally while implementing the class solution, naming should be refactored to better align with the backend message system

### PR DESCRIPTION
**Pull request by mfish33**
For some reason, Github closed this issue while rebasing to get rid of the commit from my other PR. This is not related to an active issue, but I saw an opportunity to try and improve the codebase. This PR is not done and needs to be cleaned up a little. Is this change something that is worth pursuing further? The advantage of this pattern is that it makes the subscribing to responses more explicit as well as cuts down on repeated code.

Closes #399
Workaround for #396

---
This change is [Reviewable](https://reviewable.io/reviews/graphiteeditor/graphite/394)

### CODE REVIEW

**Comment by Keavon** [comment]
For future reference, you can reopen a closed PR or issue instead of creating a new one. But no worries, we'll use this one from now on.

File: frontend/src/utilities/js-messages.ts
Comment (mfish33) [inline frontend/src/utilities/js-messages.ts:177]: Do you know why it is this way?

Comment (Keavon) [reply frontend/src/utilities/js-messages.ts:177]: It looks like `editor\src\frontend\frontend_message_handler.rs:18` defines this as `UpdateLayer { path: Path, data: LayerPanelEntry },` and that `LayerPanelEntry` also has a `path` field. I am not sure precisely what the difference is without further digging, but I'd suggest leaving it alone to prevent a regression, unless you believe that was already a mistake. We can as TrueDoctor for his thoughts, if so.

Comment (TrueDoctor) [reply frontend/src/utilities/js-messages.ts:177]: Currently they are the same thing, but I am actually not sure if that will always be the case.
It is basically a key value pair in which the value also contains the key. But that does not currently make a difference.

Comment (Keavon) [reply frontend/src/utilities/js-messages.ts:177]: Is one of the two a safer option to use? @TrueDoctor

Comment (TrueDoctor) [reply frontend/src/utilities/js-messages.ts:177]: Too late rn will try not to forget to answer tororow

Comment (TrueDoctor) [reply frontend/src/utilities/js-messages.ts:177]: I don't think there is a good reason to keep both of them in there, we should just get rid of one of them

Comment (mfish33) [reply frontend/src/utilities/js-messages.ts]: Ok, I removed path from the UpdateLayer Message

File: frontend/src/components/panels/Document.vue
Comment (mfish33) [inline frontend/src/components/panels/Document.vue:323]: What is the point of this if statement? As long as updateData is an object this will return true. The old system as well as this one should not emit a message unless it is nonnull.

Comment (Keavon) [reply frontend/src/components/panels/Document.vue:323]: To satisfy typescript, if I recall correctly. If removing it works, to ahead.

**Comment by Keavon** [comment]
Looks good. Still needs a rebase though. After rebasing, do you want me to review it?

**Comment by mfish33** [comment]
Ready for review

File: frontend/src/components/panels/Document.vue
Comment (Keavon) [inline frontend/src/components/panels/Document.vue:336]: Would you mind renaming the arrow function's parameter as the camelCased version of its name name, e.g. `updateCanvas` for this one, throughout?

File: frontend/src/components/panels/LayerTree.vue
Comment (Keavon) [inline frontend/src/components/panels/LayerTree.vue:434]: Did the autoformatter do this? I try to keep consistency with the if and else cases, so they either both have the curlies or both don't. If you have control over it, and the autoformatter doesn't change it back, could you switch it so both cases either do or don't have them? (They are arguably kinda long lines, so I don't have a strong preference either way if we use or omit curlies.)

File: frontend/src/utilities/js-message-dispatcher.ts
Comment (Keavon) [inline frontend/src/utilities/js-message-dispatcher.ts:92]: I'm thinking it might be a clearer name to just call this `registerJsMessage`, and would also reduce some code length in many places. Thoughts?

File: frontend/src/utilities/js-messages.ts
Comment (Keavon) [inline frontend/src/utilities/js-messages.ts:29]: `toRgba` would be a more accurate name.

File: frontend/src/components/widgets/inputs/SwatchPairInput.vue
Comment (Keavon) [inline frontend/src/components/widgets/inputs/SwatchPairInput.vue:146]: Might as well group this line together with the `this.primaryColor = primary.toRgb();` line above.

File: frontend/src/utilities/js-message-dispatcher.ts
Comment (Keavon) [reply frontend/src/utilities/js-message-dispatcher.ts:92]: Or would `subscribe` be a better word than `register` perhaps?

File: frontend/src/main.ts
Comment (Keavon) [inline frontend/src/main.ts:1]: Could you please write a comment describing why this has to be imported in `main.ts` if it's not used in this file?

File: frontend/src/utilities/js-message-dispatcher.ts
Comment (mfish33) [reply frontend/src/utilities/js-message-dispatcher.ts:92]: I like subscribe

File: frontend/src/utilities/js-message-handler-binding.ts
Comment (Keavon) [inline frontend/src/utilities/js-message-handler-binding.ts:4]: This file should be called `js-message-dispatcher-binding.ts` (or maybe `js-message-dispatcher-reexport.ts` if you think that's more clear) since it's meant to shadow the name of the "real" file (the reexported one, `js-message-dispatcher.ts`).

File: frontend/src/utilities/js-message-handler-binding.ts
Comment (Keavon) [inline frontend/src/utilities/js-message-handler-binding.ts:2]: Update the name in this comment too, please.

File: frontend/src/components/panels/LayerTree.vue
Comment (mfish33) [reply frontend/src/components/panels/LayerTree.vue:434]: Im an always curly bracer :)

File: frontend/src/utilities/js-messages.ts
Comment (Keavon) [inline frontend/src/utilities/js-messages.ts:33]: `toRgbaCSS()` might be slightly preferable to imply what kind of string it is: a CSS RGBA string.

File: frontend/src/utilities/js-messages.ts
Comment (Keavon) [inline frontend/src/utilities/js-messages.ts:160]: Please delete this line.

File: frontend/src/utilities/js-messages.ts
Comment (Keavon) [inline frontend/src/utilities/js-messages.ts:215]: Do you think this would, for consistency, be a good place for a transformer?

File: frontend/src/utilities/js-messages.ts
Comment (Keavon) [inline frontend/src/utilities/js-messages.ts:8]: What does `responseMarker` do? Please comment its purpose, thanks.

File: frontend/src/utilities/js-messages.ts
Comment (mfish33) [reply frontend/src/utilities/js-messages.ts:215]: I was going back and forth on that one. I can put it inside the transformer

File: frontend/src/utilities/js-message-dispatcher.ts
Comment (Keavon) [inline frontend/src/utilities/js-message-dispatcher.ts:4]: Nitpick: please put a separator newline between the first two imports (of libraries) and the other imports (of source code files).

File: frontend/src/utilities/js-messages.ts
Comment (mfish33) [reply frontend/src/utilities/js-messages.ts:8]: added the comment, "The marker provides a way to check if an object is a sub-class constructor for a jsMessage."

File: frontend/src/utilities/js-message-dispatcher.ts
Comment (Keavon) [inline frontend/src/utilities/js-message-dispatcher.ts:68]: Do you think it's correct to use PascalCase here? The variable *contains* a class, but the variable *itself* is not a class. I don't have much of an opinion on the matter, but I thought I'd bring it up to ask yours.

Comment (mfish33) [reply frontend/src/utilities/js-message-dispatcher.ts:68]: I could go either way as well. The more I look at it the more I think it should be camelCase

**Review by Keavon — APPROVED** [review]

**Comment by Keavon** [comment]
This is a rockstar PR. You know TS way better than me so I really, really appreciate your simplification of my janky code. Ready to merge after CI is done!

**Comment by Keavon** [comment]
Actually, before you merge, could you really quickly file an issue describing this change and assign it to the current sprint, then edit this PR's description to say it closes that issue number? I want it to count as one of the big features of this sprint and that's how we keep track of the sprint accomplishments (versus small commits for little fixes that aren't tracked in the task board).

**Comment by mfish33** [comment]
> This is a rockstar PR. You know TS way better than me so I really, really appreciate your simplification of my janky code. Ready to merge after CI is done!

Thank you for the kind words

**Comment by mfish33** [comment]
oops

**Comment by Keavon** [comment]
> oops

Ah, did you hit Close instead of Merge? I'm surprised reopening it requires another CI run. Kind of annoying, oh well.

**Comment by Keavon** [comment]
Haha, that's kind of funky: two sets of each check passed, since it had to rerun them on the same commit.

**Pull request merged**
